Add fly clear-versions command

clarafu opened the pull request:
## Changes proposed by this PR

closes #5418

I added a fly command that will clear the version history for a resource or a resource type. It will print out a warning message before it executes the deletion in order to show the user which other resources/resource-types are affected if they are sharing the same version history.

Only admin users will be able to use this command because of how it can be used to delete version histories of resources/resource-types across pipelines/teams. 

Through adding this fly command, I had to add 4 API endpoints. 2 are for listing out the resources/resource-types that share the same `resource_config_scope_id` (meaning that they are sharing the same version history). The other 2 endpoints are for clearing the version history for the resource or resource type.

* [x] done
* [ ] todo

## Release Note

* Can be used to clear version history of a resource or resource type
* Can only be used by an admin user
* If you have `global-resources` enabled, it can possibly delete version histories of other resources/resource-types in other pipelines so there is a warning message that will show any resources or resource types that are affected.

clarafu committed:
implements API clear resource/resource-type versions

I created two endpoints, one for clearing resource versions and the
other for clearing resource type versions. They both do the same thing
where if an admin runs the endpoint, it will trust and clear the
resource/resource type versions through it's scope. This means that if
the resource is sharing the scope with other resources, it will also
clear the versions for the other resources as well. There will be a
validation phase in the fly command that I will implement in the next
commit that will tell the user which resources will be affected by this
clearing of versions and if the user is sure they want to continue with
the clearing of versions.

[#5418]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
add API for shared resources/resource types

Added two endpoints for fetching the resources or resource types that
are sharing the same resource config scope id as the one that is passed
into the endpoint. There are two endpoints because one is for finding
the shared resources/resource types of a resource and the other is for a
resource type. The endpoints are basically the exact same though.

These endpoints are added for the fly clear-versions command that will
be added because it will be used to display to the user what
resources/resource types will be affected by the clearing of versions.

[#5418]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
clear versions endpoints to returns deleted num

Modify the clear versions for resource and resource type to return the
number of versions that were deleted. This will be shown in the fly
command after you have deleted versions using the clear versions fly
command.

[#5418]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
add fly clear-versions command

The clear-versions command can take in either a resource or a resource
type to clear the versions for. It will also print a warning message
with all the resources/resource types that the clearing of versions will
affect.

Also added the endpoints to the go-concourse package so that they can be
used by fly.

[#5418]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
fix clear versions func signature in go-concourse

thanks rui!

[#5418]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu commented on go-concourse/concourse/team.go:
```diff
@@ -51,8 +51,12 @@ type Team interface {
 	ResourceTypes(pipelineRef atc.PipelineRef) (atc.ResourceTypes, bool, error)
 	ResourceVersions(pipelineRef atc.PipelineRef, resourceName string, page Page, filter atc.Version) ([]atc.ResourceVersion, Pagination, bool, error)
+	ClearResourceVersions(pipelineRef atc.PipelineRef, resourceName string) (int64, error)
+	ClearResourceTypeVersions(pipelineRef atc.PipelineRef, resourceName string) (int64, error)
```
Great catch! I made the edits :D

xtremerui commented on atc/api/resourceserver/list_shared_for_resource.go:
```diff
@@ -0,0 +1,45 @@
+		err = json.NewEncoder(w).Encode(shared)
+		if err != nil {
+			logger.Error("failed-to-encode-shared-resources-and-types", err)
+			w.WriteHeader(http.StatusInternalServerError)
```
somewhere it returns error in the response body so fly cmd will show the error together with status code, which I think might be helpful. Have you considered to do the same for these new endpoints?

https://github.com/concourse/concourse/blob/c0e0782bfac8f345820c9a1631631fa03129e322/atc/api/resourceserver/check.go#L64

clarafu replied:
Thats a good idea, let me add that!

xtremerui replied:
@clarafu are you going to add the error mesage to the response still?

clarafu committed:
atc: write error to stdout when api endpoint fails

Addresses review comment to write the error to stdout when it fails with
500 in the api endpoints. I only added it to the db methods that have a
higher chance of erroring vs the other db methods that is only for
information gathering.

[#5418]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
fix silly mistake

[#5418]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
regenerate go-concourse team fakes

[#5418]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu commented:
@xtremerui Should be ready to be re-reviewed! :)

xtremerui reviewed (approved):
Manually run the concourse and build fly to test the new command. Works good for me!
